shadowsocks/tcp.go

```diff
@@ -62,51 +44,56 @@ func remoteIP(conn net.Conn) net.IP {
 	return nil
 }
 
-func findAccessKey(clientConn onet.DuplexConn, cipherList CipherList) (string, onet.DuplexConn, error) {
-	// This must have enough space to hold the salt + 2 bytes chunk length + AEAD tag (Oeverhead) for any cipher
-	replayBytes := make([]byte, 0, 32+2+16)
+func (s *tcpService) findAccessKey(clientConn onet.DuplexConn) (string, onet.DuplexConn, error) {
+	// All supported ciphers must use a 32-byte salt.
+	salt := [32]byte{}
+	// The ciphertext consists of a 2-byte chunk length and a 16-byte AEAD tag.
+	cipherText := [2 + 16]byte{}
 	// Constant of zeroes to use as the start chunk count. This must be as big as the max NonceSize() across all ciphers.
-	zeroCountBuf := make([]byte, 12) // MaxCountSize
+	zeroCountBuf := [12]byte{} // MaxCountSize
 	// To hold the decrypted chunk length.
 	chunkLenBuf := [2]byte{}
-	var err error
 
 	clientIP := remoteIP(clientConn)
+	if _, err := io.ReadFull(clientConn, salt[:]); err != nil {
+		logger.Debugf("TCP: Failed to read salt: %v", err)
+		return "", clientConn, err
+	}
+	if _, err := io.ReadFull(clientConn, cipherText[:]); err != nil {
+		logger.Debugf("TCP: Failed to read ciphertext: %v", err)
+		return "", clientConn, err
+	}
+
 	// Try each cipher until we find one that authenticates successfully. This assumes that all ciphers are AEAD.
 	// We snapshot the list because it may be modified while we use it.
 	// TODO: Ban and log client IPs with too many failures too quick to protect against DoS.
-	for ci, entry := range cipherList.SafeSnapshotForClientIP(clientIP) {
+	for ci, entry := range s.ciphers.SafeSnapshotForClientIP(clientIP) {
 		id, cipher := entry.Value.(*CipherEntry).ID, entry.Value.(*CipherEntry).Cipher
-		replayBytes, err = ensureBytes(clientConn, replayBytes, cipher.SaltSize())
-		if err != nil {
-			if logger.IsEnabledFor(logging.DEBUG) {
-				logger.Debugf("TCP: Failed to read salt %v: %v", id, err)
-			}
+		if cipher.SaltSize() != 32 {
+			logger.Errorf("TCP %v: Salt size must be 32, not %d", id, cipher.SaltSize())
 			continue
 		}
-		salt := replayBytes[:cipher.SaltSize()]
-		aead, err := cipher.Decrypter(salt)
-		replayBytes, err = ensureBytes(clientConn, replayBytes, cipher.SaltSize()+2+aead.Overhead())
+		aead, err := cipher.Decrypter(salt[:])
 		if err != nil {
-			if logger.IsEnabledFor(logging.DEBUG) {
-				logger.Debugf("TCP: Failed to read length %v: %v", id, err)
-			}
+			logger.Errorf("TCP %v: Decrypter initialization failed: %v", id, err)
 			continue
 		}
-		cipherText := replayBytes[cipher.SaltSize() : cipher.SaltSize()+2+aead.Overhead()]
-		_, err = aead.Open(chunkLenBuf[:0], zeroCountBuf[:aead.NonceSize()], cipherText, nil)
+		_, err = aead.Open(chunkLenBuf[:0], zeroCountBuf[:aead.NonceSize()], cipherText[:], nil)
 		if err != nil {
-			if logger.IsEnabledFor(logging.DEBUG) {
```

Comment:
Please restore all the IsEnabledFor calls. That was added on purpose, to avoid lots of unnecessary allocations.

Reply:
Done.  Note that `logger.Debugf` (etc.) all have such a check before they make any use of their arguments, so it's not obvious why this `if` statement should have any effect on the number of allocations.

Reply:
I'm open to removing the `ifs` if we can show it doesn't add allocations in the benchmark.